{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from clusterlogs import pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_colwidth', -1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('data/error0101.txt')\n",
    "#df = pd.read_csv('data/error0102.txt')\n",
    "#df = pd.read_csv('data/error0115.txt')\n",
    "#df = pd.read_csv('data/error0201.txt')\n",
    "#df = pd.read_csv('data/error0225.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>error_message</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>globus_ftp_client: the server responded with an error 500 500-Command failed : System error in stat: No such file or directory  500-A system call failed: No such file or directory  500 End.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Error on XrdCl::CopyProcess::Run(): [ERROR] Server responded with an error: [3010] Write permission denied</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>globus_ftp_client: the server responded with an error 500 500-Command failed : System error in stat: No such file or directory  500-A system call failed: No such file or directory  500 End.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>globus_ftp_client: the server responded with an error 500 Command failed : stat64 error: No such file or directory</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>globus_ftp_client: the server responded with an error 500 500-Command failed : System error in stat: No such file or directory  500-A system call failed: No such file or directory  500 End.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                                      error_message\n",
       "0  globus_ftp_client: the server responded with an error 500 500-Command failed : System error in stat: No such file or directory  500-A system call failed: No such file or directory  500 End.   \n",
       "1  Error on XrdCl::CopyProcess::Run(): [ERROR] Server responded with an error: [3010] Write permission denied                                                                                      \n",
       "2  globus_ftp_client: the server responded with an error 500 500-Command failed : System error in stat: No such file or directory  500-A system call failed: No such file or directory  500 End.   \n",
       "3  globus_ftp_client: the server responded with an error 500 Command failed : stat64 error: No such file or directory                                                                              \n",
       "4  globus_ftp_client: the server responded with an error 500 500-Command failed : System error in stat: No such file or directory  500-A system call failed: No such file or directory  500 End.   "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(703546, 1)"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster = pipeline.Chain(df, target='error_message', mode='create')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tokenization finished\n",
      "Found 21755 equal groups\n",
      "Vectorization of tokens finished\n",
      "Vectorization of sentences is finished\n",
      "Vocabulary size = 18847\n",
      "Number of dimensions is 137\n",
      "DBSCAN finished with 10045 clusters\n",
      "Matching Clusterization!\n",
      "Postprocessed with 371 clusters\n"
     ]
    }
   ],
   "source": [
    "cluster.process()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pattern</th>\n",
       "      <th>indices</th>\n",
       "      <th>cluster_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>err: Communication error on send, err: [SE][][] httpg:/:/srm/: CGSI-gSOAP running on reports Error reading token data header: Connection</td>\n",
       "      <td>[5317, 5375, 5449, 5515, 5569, 5623, 5654, 5678, 5705, 5722, 5738, 5794, 5819, 5910, 5917, 5930, 5942, 5951, 5968, 5992, 6007, 6056, 6073, 6192, 6235, 6281, 6286, 6325, 6354, 6363, 6403, 6409, 6422, 6484, 6514, 6525, 6555, 6599, 6626, 6639, 6728, 6746, 6759, 6779, 6800, 6811, 6896, 6906, 6910, 6925, 6944, 6953, 7012, 7042, 7101, 7102, 7110, 7118, 7131, 7237, 7300, 7324, 7397, 7446, 7620, 8012, 8614, 8694, 8997, 9008, 9627, 9633, 10031, 10208, 10703, 10713, 11007, 11042, 11088, 11181, 11219, 11249, 11423, 11563, 11575, 11613, 11700, 11749, 11844, 11848, 12320, 12607, 12615, 12629, 12638, 12875, 13080, 13099, 13106, 13208, ...]</td>\n",
       "      <td>79615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: []</td>\n",
       "      <td>[1, 18, 83, 93, 110, 133, 184, 274, 287, 289, 297, 304, 359, 360, 394, 397, 400, 406, 429, 434, 439, 442, 450, 457, 461, 463, 465, 466, 481, 490, 495, 511, 525, 527, 533, 535, 549, 573, 577, 579, 591, 592, 622, 641, 670, 674, 694, 715, 722, 726, 763, 770, 809, 841, 844, 875, 879, 887, 888, 901, 909, 911, 937, 939, 960, 965, 970, 980, 1004, 1010, 1020, 1048, 1054, 1095, 1098, 1122, 1138, 1144, 1145, 1167, 1170, 1175, 1185, 1217, 1240, 1257, 1272, 1273, 1297, 1300, 1316, 1331, 1337, 1346, 1351, 1355, 1361, 1373, 1380, 1411, ...]</td>\n",
       "      <td>60088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>42</td>\n",
       "      <td>Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [4003] Authentication to:1094 failed; all protocols have been tried.</td>\n",
       "      <td>[530, 633, 738, 1299, 1531, 1940, 2274, 2765, 3242, 3432, 4308, 4448, 4509, 4830, 4923, 5002, 5030, 5075, 5264, 5380, 5704, 5762, 5767, 6096, 6462, 6773, 8007, 9035, 9232, 9554, 12535, 12551, 12640, 13004, 13352, 13552, 13597, 14139, 14801, 14845, 14910, 14932, 15146, 16677, 16810, 16845, 18054, 19013, 19029, 20185, 20238, 21101, 21157, 21197, 21275, 21348, 21577, 22144, 22213, 22340, 22698, 23469, 23537, 23961, 24895, 25801, 28081, 28189, 28315, 28898, 29193, 29510, 32132, 32724, 32819, 33649, 33861, 34364, 35369, 38373, 39439, 39476, 40175, 40281, 40660, 42306, 42506, 44848, 46223, 46356, 46976, 47210, 47397, 47793, 49352, 50647, 50872, 50994, 51191, 51218, ...]</td>\n",
       "      <td>29732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>: the server responded with an error 500 error in: No such file or directory 500-A system call failed: No such file or directory 500 End.</td>\n",
       "      <td>[0, 2, 4, 5, 8, 10, 16, 17, 28, 46, 47, 54, 62, 63, 64, 66, 69, 71, 90, 104, 105, 138, 144, 159, 167, 170, 171, 172, 187, 192, 195, 196, 198, 200, 205, 206, 212, 221, 222, 223, 231, 232, 235, 237, 238, 251, 254, 257, 285, 312, 313, 335, 336, 349, 354, 361, 362, 384, 387, 388, 389, 390, 396, 398, 399, 410, 411, 413, 414, 416, 419, 426, 435, 451, 480, 493, 512, 518, 519, 588, 597, 598, 660, 662, 666, 672, 681, 687, 688, 695, 699, 702, 703, 705, 706, 713, 716, 717, 719, 730, ...]</td>\n",
       "      <td>29279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Destination file exists and overwrite is not enabled</td>\n",
       "      <td>[43, 81, 129, 255, 293, 295, 296, 498, 510, 517, 522, 712, 890, 967, 1032, 1068, 1121, 1259, 1323, 1347, 1396, 1431, 1580, 1589, 1614, 1636, 1676, 1746, 1789, 2084, 2088, 2091, 2094, 2098, 2177, 2197, 2248, 2254, 2282, 2295, 2410, 2412, 2449, 2506, 2517, 2545, 2588, 2646, 2749, 2753, 2756, 2776, 2787, 2804, 2828, 2915, 2932, 2965, 3039, 3106, 3217, 3557, 3567, 3648, 3662, 3683, 3696, 3754, 3765, 3828, 3926, 3970, 4174, 4258, 4270, 4273, 4354, 4358, 4376, 4492, 4555, 4600, 4672, 4701, 4710, 4714, 4754, 4770, 4852, 4872, 4876, 4897, 4917, 4918, 5015, 5072, 5093, 5094, 5105, 5124, ...]</td>\n",
       "      <td>26537</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                           pattern  \\\n",
       "1   err: Communication error on send, err: [SE][][] httpg:/:/srm/: CGSI-gSOAP running on reports Error reading token data header: Connection         \n",
       "0   Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: []                                                                     \n",
       "42  Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [4003] Authentication to:1094 failed; all protocols have been tried.   \n",
       "4   : the server responded with an error 500 error in: No such file or directory 500-A system call failed: No such file or directory 500 End.        \n",
       "2   Destination file exists and overwrite is not enabled                                                                                             \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             indices  \\\n",
       "1   [5317, 5375, 5449, 5515, 5569, 5623, 5654, 5678, 5705, 5722, 5738, 5794, 5819, 5910, 5917, 5930, 5942, 5951, 5968, 5992, 6007, 6056, 6073, 6192, 6235, 6281, 6286, 6325, 6354, 6363, 6403, 6409, 6422, 6484, 6514, 6525, 6555, 6599, 6626, 6639, 6728, 6746, 6759, 6779, 6800, 6811, 6896, 6906, 6910, 6925, 6944, 6953, 7012, 7042, 7101, 7102, 7110, 7118, 7131, 7237, 7300, 7324, 7397, 7446, 7620, 8012, 8614, 8694, 8997, 9008, 9627, 9633, 10031, 10208, 10703, 10713, 11007, 11042, 11088, 11181, 11219, 11249, 11423, 11563, 11575, 11613, 11700, 11749, 11844, 11848, 12320, 12607, 12615, 12629, 12638, 12875, 13080, 13099, 13106, 13208, ...]                                          \n",
       "0   [1, 18, 83, 93, 110, 133, 184, 274, 287, 289, 297, 304, 359, 360, 394, 397, 400, 406, 429, 434, 439, 442, 450, 457, 461, 463, 465, 466, 481, 490, 495, 511, 525, 527, 533, 535, 549, 573, 577, 579, 591, 592, 622, 641, 670, 674, 694, 715, 722, 726, 763, 770, 809, 841, 844, 875, 879, 887, 888, 901, 909, 911, 937, 939, 960, 965, 970, 980, 1004, 1010, 1020, 1048, 1054, 1095, 1098, 1122, 1138, 1144, 1145, 1167, 1170, 1175, 1185, 1217, 1240, 1257, 1272, 1273, 1297, 1300, 1316, 1331, 1337, 1346, 1351, 1355, 1361, 1373, 1380, 1411, ...]                                                                                                                                               \n",
       "42  [530, 633, 738, 1299, 1531, 1940, 2274, 2765, 3242, 3432, 4308, 4448, 4509, 4830, 4923, 5002, 5030, 5075, 5264, 5380, 5704, 5762, 5767, 6096, 6462, 6773, 8007, 9035, 9232, 9554, 12535, 12551, 12640, 13004, 13352, 13552, 13597, 14139, 14801, 14845, 14910, 14932, 15146, 16677, 16810, 16845, 18054, 19013, 19029, 20185, 20238, 21101, 21157, 21197, 21275, 21348, 21577, 22144, 22213, 22340, 22698, 23469, 23537, 23961, 24895, 25801, 28081, 28189, 28315, 28898, 29193, 29510, 32132, 32724, 32819, 33649, 33861, 34364, 35369, 38373, 39439, 39476, 40175, 40281, 40660, 42306, 42506, 44848, 46223, 46356, 46976, 47210, 47397, 47793, 49352, 50647, 50872, 50994, 51191, 51218, ...]   \n",
       "4   [0, 2, 4, 5, 8, 10, 16, 17, 28, 46, 47, 54, 62, 63, 64, 66, 69, 71, 90, 104, 105, 138, 144, 159, 167, 170, 171, 172, 187, 192, 195, 196, 198, 200, 205, 206, 212, 221, 222, 223, 231, 232, 235, 237, 238, 251, 254, 257, 285, 312, 313, 335, 336, 349, 354, 361, 362, 384, 387, 388, 389, 390, 396, 398, 399, 410, 411, 413, 414, 416, 419, 426, 435, 451, 480, 493, 512, 518, 519, 588, 597, 598, 660, 662, 666, 672, 681, 687, 688, 695, 699, 702, 703, 705, 706, 713, 716, 717, 719, 730, ...]                                                                                                                                                                                                  \n",
       "2   [43, 81, 129, 255, 293, 295, 296, 498, 510, 517, 522, 712, 890, 967, 1032, 1068, 1121, 1259, 1323, 1347, 1396, 1431, 1580, 1589, 1614, 1636, 1676, 1746, 1789, 2084, 2088, 2091, 2094, 2098, 2177, 2197, 2248, 2254, 2282, 2295, 2410, 2412, 2449, 2506, 2517, 2545, 2588, 2646, 2749, 2753, 2756, 2776, 2787, 2804, 2828, 2915, 2932, 2965, 3039, 3106, 3217, 3557, 3567, 3648, 3662, 3683, 3696, 3754, 3765, 3828, 3926, 3970, 4174, 4258, 4270, 4273, 4354, 4358, 4376, 4492, 4555, 4600, 4672, 4701, 4710, 4714, 4754, 4770, 4852, 4872, 4876, 4897, 4917, 4918, 5015, 5072, 5093, 5094, 5105, 5124, ...]                                                                                      \n",
       "\n",
       "    cluster_size  \n",
       "1   79615         \n",
       "0   60088         \n",
       "42  29732         \n",
       "4   29279         \n",
       "2   26537         "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#some of the clusters\n",
    "cluster.result.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Destination file exists and overwrite is not enabled'\n",
      " 'Destination file exists and overwrite is not enabled'\n",
      " 'Destination file exists and overwrite is not enabled' ...\n",
      " 'Destination file exists and overwrite is not enabled'\n",
      " 'Destination file exists and overwrite is not enabled'\n",
      " 'Destination file exists and overwrite is not enabled'] \n",
      "\n"
     ]
    }
   ],
   "source": [
    "#some messages from the cluster 2\n",
    "print(cluster.in_cluster(cluster.result, 2),\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cluster.result['pattern'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pattern</th>\n",
       "      <th>indices</th>\n",
       "      <th>cluster_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>340</td>\n",
       "      <td>DESTINATION SRM_PUT_TURL error on the turl request : [SE][StatusOfPutRequest][SRM_INVALID_PATH] Parent /gpfs/ddn/srm/cms/store/data/////000/ doesnt｟*｠!</td>\n",
       "      <td>[225297, 180602]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>329</td>\n",
       "      <td>TRANSFER ERROR: Copy failed with mode 3rd push, with error: Transfer failed: failure: Remote copy failed with status code 0: Failed to connect to｟*｠ 443:</td>\n",
       "      <td>[539911, 440066]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>364</td>\n",
       "      <td>TRANSFER globus_ftp_client: the server responded with an error｟*｠ General problem: Failed to connect [2406:f00:9:]::</td>\n",
       "      <td>[142888, 275693]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>318</td>\n",
       "      <td>DESTINATION EXISTS Destination already｟*｠ gsiftp:/eoscmsftp.cern.ch/eos/cms/store/mc///GEN/106X_upgrade2018_realistic_v4-v1//, Cancel</td>\n",
       "      <td>[424553, 316501]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>315</td>\n",
       "      <td>DESTINATION SRM_PUTDONE Error on the surl srm:/se.cis.gov.pl/dpm/cis.gov.pl/home/cms/store/data/Run2017D/JetHT/AOD/09Aug2019_UL2017-v1// while putdone : [SE][PutDone][SRM_FAILURE] Incompatible with current file status</td>\n",
       "      <td>[120527, 218340]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>311</td>\n",
       "      <td>DESTINATION SRM_PUTDONE Error on the surl srm:////rucio/mc16_13TeV///. while putdone : [SE][PutDone][SRM_INTERNAL_ERROR] Request to [&gt;PnfsManager@local] timed out.</td>\n",
       "      <td>[298977, 93086]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>362</td>\n",
       "      <td>TRANSFER globus_ftp_client: the server responded with an error｟*｠ General problem: Failed to connect [2001:660:5009:9:193:48:99:]:: Connection timed out</td>\n",
       "      <td>[687322, 243397]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>343</td>\n",
       "      <td>DESTINATION SRM_PUT_TURL error on the turl request : [SE][StatusOfPutRequest][] error accessing belle2grid4.cc.ntu.edu.tw://belle/2020-01-02: out</td>\n",
       "      <td>[372460, 112982]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>366</td>\n",
       "      <td>TRANSFER globus_ftp_client: the server responded with an error｟*｠ 550-GlobusError: v=1 c=PERMISSION_DENIED 550-GridFTP-Errno: 13 550-GridFTP-Reason: System error in mkdir (host=, user=cmsphedx, path=/store/generator/Summer13/GluGluToHHToTauTauBB_M-125_14TeV-powheg-pythia6/GEN/UpgrdPhase1Age5H_STAR17_61_V5A-v1/00000) 550-GridFTP-Error-String: Permission denied｟*｠ End.</td>\n",
       "      <td>[256947, 501180]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>307</td>\n",
       "      <td>Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [｟*｠] [ERROR] Server responded with an error: [｟*｠] Unable to read /pnfs/usatlas.bnl.gov/atlasscratchdisk/rucio/tests///; function not｟*｠</td>\n",
       "      <td>[292245, 251640]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                                                                                                                                                                                                                               pattern  \\\n",
       "340  DESTINATION SRM_PUT_TURL error on the turl request : [SE][StatusOfPutRequest][SRM_INVALID_PATH] Parent /gpfs/ddn/srm/cms/store/data/////000/ doesnt｟*｠!                                                                                                                                                                                                                             \n",
       "329  TRANSFER ERROR: Copy failed with mode 3rd push, with error: Transfer failed: failure: Remote copy failed with status code 0: Failed to connect to｟*｠ 443:                                                                                                                                                                                                                           \n",
       "364  TRANSFER globus_ftp_client: the server responded with an error｟*｠ General problem: Failed to connect [2406:f00:9:]::                                                                                                                                                                                                                                                                \n",
       "318  DESTINATION EXISTS Destination already｟*｠ gsiftp:/eoscmsftp.cern.ch/eos/cms/store/mc///GEN/106X_upgrade2018_realistic_v4-v1//, Cancel                                                                                                                                                                                                                                               \n",
       "315  DESTINATION SRM_PUTDONE Error on the surl srm:/se.cis.gov.pl/dpm/cis.gov.pl/home/cms/store/data/Run2017D/JetHT/AOD/09Aug2019_UL2017-v1// while putdone : [SE][PutDone][SRM_FAILURE] Incompatible with current file status                                                                                                                                                           \n",
       "311  DESTINATION SRM_PUTDONE Error on the surl srm:////rucio/mc16_13TeV///. while putdone : [SE][PutDone][SRM_INTERNAL_ERROR] Request to [>PnfsManager@local] timed out.                                                                                                                                                                                                                 \n",
       "362  TRANSFER globus_ftp_client: the server responded with an error｟*｠ General problem: Failed to connect [2001:660:5009:9:193:48:99:]:: Connection timed out                                                                                                                                                                                                                            \n",
       "343  DESTINATION SRM_PUT_TURL error on the turl request : [SE][StatusOfPutRequest][] error accessing belle2grid4.cc.ntu.edu.tw://belle/2020-01-02: out                                                                                                                                                                                                                                   \n",
       "366  TRANSFER globus_ftp_client: the server responded with an error｟*｠ 550-GlobusError: v=1 c=PERMISSION_DENIED 550-GridFTP-Errno: 13 550-GridFTP-Reason: System error in mkdir (host=, user=cmsphedx, path=/store/generator/Summer13/GluGluToHHToTauTauBB_M-125_14TeV-powheg-pythia6/GEN/UpgrdPhase1Age5H_STAR17_61_V5A-v1/00000) 550-GridFTP-Error-String: Permission denied｟*｠ End.   \n",
       "307  Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [｟*｠] [ERROR] Server responded with an error: [｟*｠] Unable to read /pnfs/usatlas.bnl.gov/atlasscratchdisk/rucio/tests///; function not｟*｠                                                                                                                                                                 \n",
       "\n",
       "              indices  cluster_size  \n",
       "340  [225297, 180602]  2             \n",
       "329  [539911, 440066]  2             \n",
       "364  [142888, 275693]  2             \n",
       "318  [424553, 316501]  2             \n",
       "315  [120527, 218340]  2             \n",
       "311  [298977, 93086]   2             \n",
       "362  [687322, 243397]  2             \n",
       "343  [372460, 112982]  2             \n",
       "366  [256947, 501180]  2             \n",
       "307  [292245, 251640]  2             "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clusters, outliers = cluster.split_clusters(cluster.result, 'cluster_size', threshold=3)\n",
    "outliers.head(10) #clusters containing fewer than 3 messages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_size</th>\n",
       "      <th>pattern</th>\n",
       "      <th>mean_similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>79615</td>\n",
       "      <td>err: Communication error on send, err: [SE][][] httpg:/:/srm/: CGSI-gSOAP running on reports Error reading token data header: Connection</td>\n",
       "      <td>0.94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>60088</td>\n",
       "      <td>Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: []</td>\n",
       "      <td>0.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>29732</td>\n",
       "      <td>Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [4003] Authentication to:1094 failed; all protocols have been tried.</td>\n",
       "      <td>0.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>29279</td>\n",
       "      <td>: the server responded with an error 500 error in: No such file or directory 500-A system call failed: No such file or directory 500 End.</td>\n",
       "      <td>0.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>26537</td>\n",
       "      <td>Destination file exists and overwrite is not enabled</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>24407</td>\n",
       "      <td>｟*｠: Unable to connect to:2811｟*｠: System error in connect: Connection｟*｠: A system call failed: Connection</td>\n",
       "      <td>0.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>19802</td>\n",
       "      <td>Error reported from srm_ifce : 2 [SE][Ls][SRM_INVALID_PATH] No such file or directory</td>\n",
       "      <td>0.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>18333</td>\n",
       "      <td>end-of-file was reached｟*｠: An end of file occurred</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>18207</td>\n",
       "      <td>[gfalt_copy_file][perform_copy][gfal_xrootd_3rd_copy][gfal_xrootd_3rd_copy_bulk] Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: []</td>\n",
       "      <td>0.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>17760</td>\n",
       "      <td>SOURCE SRM_GET_TURL error on the turl request : [SE][][SRM_FILE_UNAVAILABLE] File is.</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>16797</td>\n",
       "      <td>TRANSFER Transfer canceled because the gsiftp performance marker timeout of 360 seconds has been exceeded, or all performance markers during that period indicated zero bytes transferred</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>15907</td>\n",
       "      <td>Protocol not supported or path/url invalid: root:/:1094/dpm//home/atlas/atlasscratchdisk/rucio/tests///</td>\n",
       "      <td>0.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>15783</td>\n",
       "      <td>err: Communication error on send, err: [SE][][] httpg:/:/srm/managerv2: CGSI-gSOAP running on reports could not open connection to:</td>\n",
       "      <td>0.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>14821</td>\n",
       "      <td>Protocol not supported or path/url invalid: root:/:1094//atlas/atlasscratchdisk/rucio/tests///</td>\n",
       "      <td>0.66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>14629</td>\n",
       "      <td>Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [] Unable to //////;</td>\n",
       "      <td>0.61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>14581</td>\n",
       "      <td>Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [3010] Unable to open /dpm/nipne.ro/home/atlas/atlasscratchdisk/rucio/tests//; no delegated credentials for tpc</td>\n",
       "      <td>0.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>13091</td>\n",
       "      <td>: the server responded with an error 421 Service busy: Connection limit exceeded. Please try again later. Closing control connection.</td>\n",
       "      <td>0.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>12665</td>\n",
       "      <td>: the server responded with an error 530 530-globus_xio: Authentication Error 530-OpenSSL Error: s3_srvr.c:: in library: SSL routines, function: certificate 530-globus_gsi_callback_module: Could not verify credential 530-globus_gsi_callback_module: The certificate: with subject: /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=/CN=/CN=/CN=/CN=</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>12409</td>\n",
       "      <td>Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [3010] Unable to create /pnfs/usatlas.bnl.gov/atlasscratchdisk/rucio/tests///; permission denied</td>\n",
       "      <td>0.85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19</td>\n",
       "      <td>11878</td>\n",
       "      <td>Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [3010] Access denied: /pnfs/triumf.ca/data/atlas/atlasscratchdisk/rucio/tests//</td>\n",
       "      <td>0.84</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cluster_size  \\\n",
       "0   79615          \n",
       "1   60088          \n",
       "2   29732          \n",
       "3   29279          \n",
       "4   26537          \n",
       "5   24407          \n",
       "6   19802          \n",
       "7   18333          \n",
       "8   18207          \n",
       "9   17760          \n",
       "10  16797          \n",
       "11  15907          \n",
       "12  15783          \n",
       "13  14821          \n",
       "14  14629          \n",
       "15  14581          \n",
       "16  13091          \n",
       "17  12665          \n",
       "18  12409          \n",
       "19  11878          \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                 pattern  \\\n",
       "0   err: Communication error on send, err: [SE][][] httpg:/:/srm/: CGSI-gSOAP running on reports Error reading token data header: Connection                                                                                                                                                                                                               \n",
       "1   Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: []                                                                                                                                                                                                                                                                           \n",
       "2   Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [4003] Authentication to:1094 failed; all protocols have been tried.                                                                                                                                                                                                         \n",
       "3   : the server responded with an error 500 error in: No such file or directory 500-A system call failed: No such file or directory 500 End.                                                                                                                                                                                                              \n",
       "4   Destination file exists and overwrite is not enabled                                                                                                                                                                                                                                                                                                   \n",
       "5   ｟*｠: Unable to connect to:2811｟*｠: System error in connect: Connection｟*｠: A system call failed: Connection                                                                                                                                                                                                                                            \n",
       "6   Error reported from srm_ifce : 2 [SE][Ls][SRM_INVALID_PATH] No such file or directory                                                                                                                                                                                                                                                                  \n",
       "7   end-of-file was reached｟*｠: An end of file occurred                                                                                                                                                                                                                                                                                                    \n",
       "8   [gfalt_copy_file][perform_copy][gfal_xrootd_3rd_copy][gfal_xrootd_3rd_copy_bulk] Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: []                                                                                                                                                                                          \n",
       "9   SOURCE SRM_GET_TURL error on the turl request : [SE][][SRM_FILE_UNAVAILABLE] File is.                                                                                                                                                                                                                                                                  \n",
       "10  TRANSFER Transfer canceled because the gsiftp performance marker timeout of 360 seconds has been exceeded, or all performance markers during that period indicated zero bytes transferred                                                                                                                                                              \n",
       "11  Protocol not supported or path/url invalid: root:/:1094/dpm//home/atlas/atlasscratchdisk/rucio/tests///                                                                                                                                                                                                                                                \n",
       "12  err: Communication error on send, err: [SE][][] httpg:/:/srm/managerv2: CGSI-gSOAP running on reports could not open connection to:                                                                                                                                                                                                                    \n",
       "13  Protocol not supported or path/url invalid: root:/:1094//atlas/atlasscratchdisk/rucio/tests///                                                                                                                                                                                                                                                         \n",
       "14  Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [] Unable to //////;                                                                                                                                                                                                                                                         \n",
       "15  Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [3010] Unable to open /dpm/nipne.ro/home/atlas/atlasscratchdisk/rucio/tests//; no delegated credentials for tpc                                                                                                                                                              \n",
       "16  : the server responded with an error 421 Service busy: Connection limit exceeded. Please try again later. Closing control connection.                                                                                                                                                                                                                  \n",
       "17  : the server responded with an error 530 530-globus_xio: Authentication Error 530-OpenSSL Error: s3_srvr.c:: in library: SSL routines, function: certificate 530-globus_gsi_callback_module: Could not verify credential 530-globus_gsi_callback_module: The certificate: with subject: /DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=/CN=/CN=/CN=/CN=   \n",
       "18  Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [3010] Unable to create /pnfs/usatlas.bnl.gov/atlasscratchdisk/rucio/tests///; permission denied                                                                                                                                                                             \n",
       "19  Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [3010] Access denied: /pnfs/triumf.ca/data/atlas/atlasscratchdisk/rucio/tests//                                                                                                                                                                                              \n",
       "\n",
       "    mean_similarity  \n",
       "0   0.94             \n",
       "1   0.99             \n",
       "2   0.90             \n",
       "3   0.98             \n",
       "4   1.00             \n",
       "5   0.89             \n",
       "6   0.99             \n",
       "7   1.00             \n",
       "8   0.99             \n",
       "9   1.00             \n",
       "10  1.00             \n",
       "11  0.68             \n",
       "12  0.84             \n",
       "13  0.66             \n",
       "14  0.61             \n",
       "15  0.86             \n",
       "16  0.98             \n",
       "17  0.96             \n",
       "18  0.85             \n",
       "19  0.84             "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#top 20 clusters by the size of cluster\n",
    "top_numb = 20\n",
    "top_clusters = cluster.validation(clusters)[['cluster_size','pattern','mean_similarity']].sort_values(by=[\"cluster_size\"], ascending=False) #clusters sorted by similarity inside the cluster\n",
    "top_clusters.head(top_numb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_size</th>\n",
       "      <th>pattern</th>\n",
       "      <th>mean_similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>14629</td>\n",
       "      <td>Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [] Unable to //////;</td>\n",
       "      <td>0.61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>14821</td>\n",
       "      <td>Protocol not supported or path/url invalid: root:/:1094//atlas/atlasscratchdisk/rucio/tests///</td>\n",
       "      <td>0.66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>15907</td>\n",
       "      <td>Protocol not supported or path/url invalid: root:/:1094/dpm//home/atlas/atlasscratchdisk/rucio/tests///</td>\n",
       "      <td>0.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>26</td>\n",
       "      <td>8445</td>\n",
       "      <td>Protocol not supported or path/url invalid: root:/:1094//atlasscratchdisk/rucio/tests///</td>\n",
       "      <td>0.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>170</td>\n",
       "      <td>39</td>\n",
       "      <td>DESTINATION SRM_PUTDONE Error on the surl srm://dpm//home/cms/store/////// while putdone : [SE][PutDone][SRM_FAILURE] Incompatible with current file status</td>\n",
       "      <td>0.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>289</td>\n",
       "      <td>3</td>\n",
       "      <td>DESTINATION SRM_PUTDONE Error on the surl srm:/se01.indiacms.res.in/dpm/indiacms.res.in/home/cms/store/user/////// while putdone : [SE][PutDone][SRM_INTERNAL_ERROR]｟*｠ out</td>\n",
       "      <td>0.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>21</td>\n",
       "      <td>11260</td>\n",
       "      <td>Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [3012] sync - TPC open failed for src_url=root:/://atlasscratchdisk/rucio/tests///?fst.readahead=true</td>\n",
       "      <td>0.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>42</td>\n",
       "      <td>3484</td>\n",
       "      <td>Protocol not supported or path/url invalid: root:/:1094/pnfs//atlasscratchdisk/rucio/tests///</td>\n",
       "      <td>0.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>292</td>\n",
       "      <td>3</td>\n",
       "      <td>TRANSFER globus_ftp_client: the server responded with an error 500 Command failed. : Failed to close file in HDFS. (host=, user=, path=/store///////)</td>\n",
       "      <td>0.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>199</td>\n",
       "      <td>16</td>\n",
       "      <td>DESTINATION MAKE_PARENT srm-ifce err: Permission denied, err: [SE][Mkdir][SRM_AUTHORIZATION_FAILURE] httpg:/:8443/srm/managerv2: srm://castor/ads.rl.ac.uk/prod///////: User can not create a directory at this level</td>\n",
       "      <td>0.74</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     cluster_size  \\\n",
       "14   14629          \n",
       "13   14821          \n",
       "11   15907          \n",
       "26   8445           \n",
       "170  39             \n",
       "289  3              \n",
       "21   11260          \n",
       "42   3484           \n",
       "292  3              \n",
       "199  16             \n",
       "\n",
       "                                                                                                                                                                                                                   pattern  \\\n",
       "14   Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [] Unable to //////;                                                                                                                          \n",
       "13   Protocol not supported or path/url invalid: root:/:1094//atlas/atlasscratchdisk/rucio/tests///                                                                                                                          \n",
       "11   Protocol not supported or path/url invalid: root:/:1094/dpm//home/atlas/atlasscratchdisk/rucio/tests///                                                                                                                 \n",
       "26   Protocol not supported or path/url invalid: root:/:1094//atlasscratchdisk/rucio/tests///                                                                                                                                \n",
       "170  DESTINATION SRM_PUTDONE Error on the surl srm://dpm//home/cms/store/////// while putdone : [SE][PutDone][SRM_FAILURE] Incompatible with current file status                                                             \n",
       "289  DESTINATION SRM_PUTDONE Error on the surl srm:/se01.indiacms.res.in/dpm/indiacms.res.in/home/cms/store/user/////// while putdone : [SE][PutDone][SRM_INTERNAL_ERROR]｟*｠ out                                             \n",
       "21   Error on XrdCl:CopyProcess:Run(): [ERROR] Server responded with an error: [3012] sync - TPC open failed for src_url=root:/://atlasscratchdisk/rucio/tests///?fst.readahead=true                                         \n",
       "42   Protocol not supported or path/url invalid: root:/:1094/pnfs//atlasscratchdisk/rucio/tests///                                                                                                                           \n",
       "292  TRANSFER globus_ftp_client: the server responded with an error 500 Command failed. : Failed to close file in HDFS. (host=, user=, path=/store///////)                                                                   \n",
       "199  DESTINATION MAKE_PARENT srm-ifce err: Permission denied, err: [SE][Mkdir][SRM_AUTHORIZATION_FAILURE] httpg:/:8443/srm/managerv2: srm://castor/ads.rl.ac.uk/prod///////: User can not create a directory at this level   \n",
       "\n",
       "     mean_similarity  \n",
       "14   0.61             \n",
       "13   0.66             \n",
       "11   0.68             \n",
       "26   0.70             \n",
       "170  0.70             \n",
       "289  0.70             \n",
       "21   0.74             \n",
       "42   0.74             \n",
       "292  0.74             \n",
       "199  0.74             "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#most inaccurate clusters\n",
    "top_clusters[['cluster_size','pattern','mean_similarity']].sort_values(by=[\"mean_similarity\"]).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'66.28% of all logs are in the top 20 of 371 clusters'"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "perc = round(top_clusters['cluster_size'].head(20).sum()/(df.shape[0]-1)*100,2)\n",
    "f\"{perc}% of all logs are in the top {top_numb} of {len(cluster.result)} clusters\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'tokenization': 314.0405,\n",
       " 'group_equals': 30.2672,\n",
       " 'tokens_vectorization': 2.7288,\n",
       " 'sentence_vectorization': 4.0955,\n",
       " 'ml_clusterization': 5.7293,\n",
       " 'matching_clusterization': 10.5776,\n",
       " 'process': 367.4536,\n",
       " 'validation': 8.3687}"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster.timings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_clusters.head(top_numb).to_csv(\"topclusters0101.txt\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
